flink-connectors/flink-connector-files/src/main/java/org/apache/flink/connector/file/table/batch/compact/BatchCompactCoordinator.java

```diff
@@ -0,0 +1,155 @@
+
+import org.apache.flink.connector.file.table.BinPacking;
+import org.apache.flink.connector.file.table.stream.compact.CompactMessages.CompactionUnit;
+import org.apache.flink.connector.file.table.stream.compact.CompactMessages.CoordinatorInput;
+import org.apache.flink.connector.file.table.stream.compact.CompactMessages.CoordinatorOutput;
+import org.apache.flink.connector.file.table.stream.compact.CompactMessages.InputFile;
+import org.apache.flink.core.fs.FileSystem;
+import org.apache.flink.core.fs.Path;
+import org.apache.flink.runtime.state.StateInitializationContext;
+import org.apache.flink.streaming.api.operators.AbstractStreamOperator;
+import org.apache.flink.streaming.api.operators.BoundedOneInput;
+import org.apache.flink.streaming.api.operators.OneInputStreamOperator;
+import org.apache.flink.streaming.runtime.streamrecord.StreamRecord;
+import org.apache.flink.util.function.SupplierWithException;
+
+import java.io.IOException;
+import java.util.ArrayList;
+import java.util.Collections;
+import java.util.HashMap;
+import java.util.List;
+import java.util.Map;
+import java.util.function.Function;
+
+/**
+ * Coordinator for compaction in batch mode. It will collect the written files in {@link
+ * BatchFileWriter} and determine whether to compact files or not as well as what files should be
+ * merged into a single file.
+ */
+public class BatchCompactCoordinator extends AbstractStreamOperator<CoordinatorOutput>
+        implements OneInputStreamOperator<CoordinatorInput, CoordinatorOutput>, BoundedOneInput {
+
+    private static final long serialVersionUID = 1L;
+
+    private final SupplierWithException<FileSystem, IOException> fsFactory;
+    private final long compactAverageSize;
+    private final long compactTargetSize;
+
+    private transient FileSystem fs;
+    // the mapping from written partitions to the corresponding files.
+    private transient Map<String, List<Path>> inputFiles;
+
+    public BatchCompactCoordinator(
+            SupplierWithException<FileSystem, IOException> fsFactory,
+            long compactAverageSize,
+            long compactTargetSize) {
+        this.fsFactory = fsFactory;
+        this.compactAverageSize = compactAverageSize;
+        this.compactTargetSize = compactTargetSize;
+    }
+
+    @Override
+    public void initializeState(StateInitializationContext context) throws Exception {
+        super.initializeState(context);
+        fs = fsFactory.get();
+        inputFiles = new HashMap<>();
+    }
+
+    @Override
+    public void processElement(StreamRecord<CoordinatorInput> element) throws Exception {
+        CoordinatorInput coordinatorInput = element.getValue();
+        if (coordinatorInput instanceof InputFile) {
+            InputFile file = (InputFile) coordinatorInput;
+            // collect the written files
+            inputFiles
+                    .computeIfAbsent(file.getPartition(), k -> new ArrayList<>())
+                    .add(file.getFile());
+        }
```

Comment:
throw an exception for other unexpected type

Reply:
done